Spack recipes for ROCm software components-Phase1

srekolam opened the pull request:
This Pull request is for the spack recipes for the AMD ROCm stage 1 build components.

srekolam committed:
Spack recipes for ROCm Stage 1 Build components

adamjstewart commented:
I see some overlap with packages added by @aweits in #17271. We should figure out an acceptable naming convention.

saadrahim commented:
Stage 1 is internal AMD nomenclature. We should rename the PR to reflect something that is relevant for all.

aweits commented:
This fails to build correctly for me, I suspect that it's the "llvm build finding the system gcc issue", will post info when I'm able.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

srekolam commented:
adamjstewart and  tcojean  , I am currently looking at the flake8 errors other than that let me know if there are any review comments that needs to be addressed.

Commit:
fix flake8 errors

aweits commented:
> This fails to build correctly for me, I suspect that it's the "llvm build finding the system gcc issue", will post info when I'm able.

The fix for my issue is to add this:
```
    def cmake_args(self):
        args = [
            '-DCMAKE_VERBOSE_MAKEFILE=1',
            '-DLLVM_ENABLE_PROJECTS=clang;lld;clang-tools-extra;compiler-rt',
            '-DLLVM_ENABLE_ASSERTIONS=1',
        ]

        if self.compiler.name == "gcc":
            gcc_prefix = ancestor(self.compiler.cc, 2)
            args.append("-DGCC_INSTALL_PREFIX=" + gcc_prefix)

        return args
```
which parallels the spack llvm package. Otherwise on my system (at runtime), llvm finds and attempts to use the older system gcc for headers/etc - this ensures it finds the gcc it was built with.

srekolam commented:
fixed the flake8 errors. 
@aweits , can you use the gcc from dev-toolset-7 for the compiling the llvm-amdgpu.

aweits commented:
> fixed the flake8 errors.
> @aweits , can you use the gcc from dev-toolset-7 for the compiling the llvm-amdgpu.

I'm currently using spack-installed gcc@8.2.0, what's dev-toolset-7?

srekolam commented:
> > fixed the flake8 errors.
> > @aweits , can you use the gcc from devtoolset-7 for the compiling the llvm-amdgpu.
> 
> I'm currently using spack-installed gcc@8.2.0, what's dev-toolset-7?

I have been using gcc versions after installing devtoolset-7-all.x86_64. This has the gcc @7.3.1.

srekolam committed:
fixes for flake8 errors

srekolam commented:
@adamjstewart , i have addressed all the flake8 errors. can you let me know if there is anything from my end for this PR

adamjstewart commented:
Internet is down at my place but I'll review when it comes back.

mpbelhorn commented:
Perhaps this is something that should be brought up as an issue on the `ROCclr` issue tracker rather than here, but I'm wondering about [these hard-coded macros](https://github.com/ROCm-Developer-Tools/ROCclr/blob/roc-3.5.0/CMakeLists.txt#L60) that enable x86_64/SSE specific types by way of [xmmintrin.h](https://github.com/ROCm-Developer-Tools/ROCclr/blob/roc-3.5.0/top.hpp#L44-L46) for non ARM architectures.

I'm trying to build ROCclr on ppc64le architecture and it's not clear in the ROCclr documentation if that's even supported. The hard-coded macros enabling x86_64 intrinsics suggest no. If there are hard architecture constraints, I would propose adding a conflict directive, perhaps something like

```
class Rocclr(CMakePackage):
    # ...
    conflicts('target=ppc64',
                  when='@:3.5.1',  # or until a version is released that is supported on non-x86_64 targets
                   msg='ROCclr requires x86 intriniscs except on ARM')
    # ...
```

though, ideally, the `ROCclr` CMakeLists wouldn't make assumptions about the architecture like this if it's possible to build it on Power ISA. I don't want to pollute this MR with the details, but I'd like to discuss the best approach to build the ROCm software stack from source on ppc64le/nvcc hardware. This includes

- what platform ("deprecated HCC" vs "ROCclr with x86 intrinics" vs "something I've missed") to target to build it successfully
- and making clear what ROCm components are needed/useful on NVCC/non-x86_64 platforms through spack platform/target `conflicts()` constraints on the appropriate packages.

At a minimum, I'd like to offer users on our ppc64le/nvcc machines `hip` and `hipcc`. And while I know the math libs generally fallback to the respective CUDA counterparts/implementations when building HIP codes on an nvcc platform, our users have expressed interest in being able to use `hipBLAS`, `rocFFT`, `rocRAND`, `hipSPARSE` on our NVCC-platform machines. I think it would be a good idea to have an optional `cuda` variant and dependency on the libraries which need CUDA headers at build-time to support NVCC platforms.

To be clear, my interest here is to help our users currently running on Nvidia hardware have a headstart porting their codes to AMD/heterogeneous hardware. While it may seem counterproductive from certain perspectives to focus on NVCC platforms, the effort is ultimately so we can migrate users to AMD hardware with as much satisfaction and as little pain as possible.

srekolam commented:
@adamjstewart , can you please review and approve this request

adamjstewart commented on var/spack/repos/builtin/packages/atmi/package.py:
```diff
@@ -0,0 +1,46 @@
+    version('3.5.0', sha256='3fb57d2e583fab82bd0582d0c2bccff059ca91122c18ac49a7770a8bb041a37b')
+
+    variant('build_type', default='Release', values=("Release", "Debug"), description='CMake build type')
+    depends_on('cmake@3.5.2', type='build')
```
```suggestion
    depends_on('cmake@2.8:', type='build')
```
There's no reason to force a particular version of CMake is there? The `CMakeLists.txt` says 2.8+ is the only requirement.

srekolam replied:
@adamjstewart , we will look into cmakeLists.txt for each recipe and modify the above . Thanks for your inputs.

srekolam replied:
@adamjstewart  ,today I tried to change as per the cmakelist.txt for each recipe but it is creating some conflicts 
Like hsakmt-roct  and hsa-rocr-dev require cmake-3.5  ; llvm-amdgpu requires cmake 3.4.3 and comgr - cmake 3.2.0;
rocm-device-libs -cmake 3.4.3 and atmi - cmake 2.8
Say for atmi, it is dependent on hsakmt-roct and hsa-rocr-dev and they require cmake 3.5 ..whereas cmakelists.txt for atmi mentions it requires cmake 2.8 .. this seems to create conflict when we try to concretize the spec. 
I am thinking i will force a particular version of cmake unless there is any solution. let me know what you think.

adamjstewart replied:
The solution is to add a colon at the end: `cmake@2.8:`. This means CMake 2.8+. If all of your packages use the ending colon, then Spack will use any version that satisfies all of them, in this case likely the newest version.

adamjstewart commented on var/spack/repos/builtin/packages/atmi/package.py:
```diff
@@ -0,0 +1,46 @@
+
+    variant('build_type', default='Release', values=("Release", "Debug"), description='CMake build type')
+    depends_on('cmake@3.5.2', type='build')
+    depends_on('comgr@3.5:', type='build', when='@3.5:')
+    depends_on('hsa-rocr-dev@3.5:', type='build', when='@3.5:')
+    depends_on('libelf@0.8:', type='build', when='@3.5:')
```
`when='@3.5:'` isn't needed here since you don't have any versions older than 3.5. Also, are you sure these are all build dependencies? Build deps are added to the `PATH` during installation, but a lot of these looks like libraries that should be linked to. The default dep type (build/link) should be sufficient, or you can explicitly specify link if it doesn't need to be in the `PATH`.

srekolam replied:
when='@3.5:' was added based on need to support future releases too. As you may find atmi for rocm-3.5.0 is dependent on rocm-3.5.0 versions of comgr, hsa-rocr-dev etc . Whereas atmi for rocm-3.8.0 will be dependent on rocm-3.8.0 versions of comgr ,hsa-rocr-dev. 
My understanding is that for rocm3.8.0, we will be adding the below in addition to what we have currently
     depends_on('comgr@3.8:', type='build', when='@3.8:')
    depends_on('hsa-rocr-dev@3.8:', type='build', when='@3.8:')
The build deps were added mainly because atmi is dependent on comgr and hsa-rocr-dev. If they are not built already, those are built initially and then atmi will be built.

adamjstewart replied:
In that case, a nice way to handle this might be:
```suggestion
    for ver in ['3.5.0', '3.8.0', ...]:
        depends_on('comgr@' + ver, type='build', when='@' + ver)
        depends_on('hsa-rocr-dev@ver', type='build', when='@' + ver)
    depends_on('libelf@0.8:', type='build')
```
Does romc 3.5.0 require exactly comgr 3.5.0, or can it build with 3.5.0+? If it's the former, the above should be correct, if it's the latter, add a trailing colon to the version number to represent +.

srekolam replied:
yes, rcom 3.5.0 require exactly comgr 3.5.0 etc

adamjstewart commented on var/spack/repos/builtin/packages/atmi/package.py:
```diff
@@ -0,0 +1,46 @@
+        # workaround for /opt/rocm/.info/version
+        spec = self.spec
+        args = [
+            '-DROCM_VERSION=3.5.0-2588',
```
Can you describe this workaround in more detail? I try to avoid hard-coding versions in packages since Spack packages generally support (or will someday support) multiple versions of the same package.

srekolam replied:
This work around requires changes and will be updated

adamjstewart commented on var/spack/repos/builtin/packages/atmi/package.py:
```diff
@@ -0,0 +1,46 @@
+        args = [
+            '-DROCM_VERSION=3.5.0-2588',
+            '-DCMAKE_VERBOSE_MAKEFILE = 1',
+            '-DCMAKE_PREFIX_PATH={}/include/hsa;{}/hsa/lib;\
+                {}/include;{}/lib;{}/include;{}/lib'.format(
+                spec['hsa-rocr-dev'].prefix, spec['hsa-rocr-dev'].prefix,
+                spec['hsakmt-roct'].prefix, spec['hsakmt-roct'].prefix,
+                spec['libelf'].prefix, spec['libelf'].prefix)
```
This isn't needed, it's already added automatically by the `CMakePackage` base class.

srekolam replied:
sure. Will remove them as per the comment,

adamjstewart commented on var/spack/repos/builtin/packages/hip/package.py:
```diff
@@ -0,0 +1,48 @@
+        args = [
+            '-DHIP_COMPILER=clang',
+            '-DHIP_PLATFORM=rocclr',
+            '-DHSA_PATH={}'.format(self.spec['hsa-rocr-dev'].prefix),
```
```suggestion
            '-DHSA_PATH={0}'.format(self.spec['hsa-rocr-dev'].prefix),
```
This is needed for Python 2.6 support

srekolam replied:
will be addressed as your your comments

arjun-raj-kuppala replied:
@adamjstewart Out of curiosity, what version of python does spack use..is it dependent on the system environment??

haampie replied:
Spack aims to support Python 2.6 and above

adamjstewart replied:
Specifically 2.6-2.7 and 3.5-3.8: https://spack.readthedocs.io/en/latest/getting_started.html

adamjstewart commented on var/spack/repos/builtin/packages/hip/package.py:
```diff
@@ -0,0 +1,48 @@
+            'hip-config.cmake.in', string=True)
+
+    def setup_build_environment(self, build_env):
+        build_env.unset('PERL5LIB')
```
Is perl needed to build and/or run this package? You don't have any dependency on it.

adamjstewart commented on var/spack/repos/builtin/packages/comgr/package.py:
```diff
@@ -0,0 +1,36 @@
+    def cmake_args(self):
+        args = [
+            '-DCMAKE_VERBOSE_MAKEFILE=1',
+            '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH=FALSE'
```
I'm not super familiar with CMake, but generally we try to avoid CMake-specific options on the package level, we like to put those in the `CMakePackage` base class so that all packages are installed in a uniform way.

adamjstewart commented on var/spack/repos/builtin/packages/rocclr/package.py:
```diff
@@ -0,0 +1,46 @@
+             expand=True,
+             destination='',
+             placement='opencl-on-vdi')
+    build_directory = '../../rocclr_build'
```
Out of curiosity, why does the build_directory need to be overwritten here?

haampie replied:
@adamjstewart: I think the idea is to create a directory _outside_ of the staging directory :sweat_smile:. And this is would be reused inside `rocm-opencl/package.py` or so... can't be correct :p See a few lines below, it seems like the idea is to make the folder persistent, and to only clean it up right before running `cmake`.

The issue is I think that when rocclr is installed, it does not install everything needed for rocm-opencl. The proper fix is to install the missing files of course. That's also what is being done in the AUR package. Ping @srekolam

adamjstewart commented on var/spack/repos/builtin/packages/rocfft/package.py:
```diff
@@ -0,0 +1,61 @@
+
+    def cmake_args(self):
+        # Finding the version of clang
+        hipcc = Executable(join_path(self.spec['hip'].prefix.bin, 'hipcc'))
+        version = hipcc('--version', output=str)
+        version_group = re.search(r"clang version (\S+)", version)
+        version_number = version_group.group(1)
```
```suggestion
        version_number = self.spec['hip'].version
```

adamjstewart commented on var/spack/repos/builtin/packages/rocfft/package.py:
```diff
@@ -0,0 +1,61 @@
+            '-DUSE_HIP_CLANG=ON',
+            '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH=FALSE',
+            '-DHIP_CLANG_INCLUDE_PATH={}/lib/clang/{}/include'.format(
+                self.spec['llvm-amdgpu'].prefix, version_number)
```
It might be easier/more reliable/more reusable to override the `headers` property of `llvm-amdgpu` and then use that here.

adamjstewart commented on var/spack/repos/builtin/packages/rocgdb/package.py:
```diff
@@ -0,0 +1,54 @@
+    depends_on('cmake@3.5.2', type='build')
+    depends_on('texinfo')
+    depends_on('bison')
+    depends_on('flex')
```
These 3 are often build dependencies

srekolam replied:
Will be updated with the type='build' for the above

adamjstewart commented on var/spack/repos/builtin/packages/rocgdb/package.py:
```diff
@@ -0,0 +1,54 @@
+    depends_on('flex')
+    depends_on('libunwind')
+    depends_on('expat')
+    depends_on('python')
```
Python is often a build/run dependency. Do you actually link to the python libraries? Is it only needed at build-time?

srekolam replied:
it was required for the build time. Will update accordingly.

adamjstewart commented on var/spack/repos/builtin/packages/rocgdb/package.py:
```diff
@@ -0,0 +1,54 @@
+            '--disable-gdbtk',
+            '--disable-shared',
+            '--with-expat',
+            '--with-system-zlib'
```
I think you're missing a zlib dependency. You shouldn't rely on the fact that other dependencies like python depend on zlib, you should always explicitly list all direct dependencies.

srekolam replied:
Will add depends_on(zlib, type='build',when='@3.5:')

adamjstewart replied:
Build dependencies only add executables to the `PATH`. Since `zlib` doesn't have any executables, I assume you want `type='link'`. This tells the compiler wrappers to inject `zlib` into the RPATH, among other things.

adamjstewart replied:
Still missing a zlib dependency here.

adamjstewart commented on var/spack/repos/builtin/packages/rocm-smi/package.py:
```diff
@@ -0,0 +1,34 @@
+        popen('cp -R {}/tests/ {}/smi-test/'.format(self.build_directory,
+                                                    prefix))
+        popen('cp -R {}/test-rocm-smi.sh {}/smi-test'.format(
+              self.build_directory, prefix))
```
Spack hopes to someday support Windows, so this won't work. Spack has builtin `copy`, `copy_tree`, `install`, `install_tree`, `mkdir`, and `mkdirp` methods than you can use instead. These are OS-independent.

srekolam replied:
will address it as per your comments

adamjstewart replied:
Still needs to be addressed

adamjstewart reviewed (changes requested):
Sorry for the delay, my internet is finally back!

I added several comments. Most of these comments apply to more than one package, but I didn't repeat myself on every package. Let me know if you have any questions, and I'll re-review once you're done addressing these.

aweits commented:
> > > fixed the flake8 errors.
> > > @aweits , can you use the gcc from devtoolset-7 for the compiling the llvm-amdgpu.
> > 
> > 
> > I'm currently using spack-installed gcc@8.2.0, what's dev-toolset-7?
> 
> I have been using gcc versions after installing devtoolset-7-all.x86_64. This has the gcc @7.3.1.

Understood. Either way, we should add that GCC_INSTALL_PREFIX.

srekolam commented:
> Perhaps this is something that should be brought up as an issue on the `ROCclr` issue tracker rather than here, but I'm wondering about [these hard-coded macros](https://github.com/ROCm-Developer-Tools/ROCclr/blob/roc-3.5.0/CMakeLists.txt#L60) that enable x86_64/SSE specific types by way of [xmmintrin.h](https://github.com/ROCm-Developer-Tools/ROCclr/blob/roc-3.5.0/top.hpp#L44-L46) for non ARM architectures.
> 
> I'm trying to build ROCclr on ppc64le architecture and it's not clear in the ROCclr documentation if that's even supported. The hard-coded macros enabling x86_64 intrinsics suggest no. If there are hard architecture constraints, I would propose adding a conflict directive, perhaps something like
> 
> ```
> class Rocclr(CMakePackage):
>     # ...
>     conflicts('target=ppc64',
>                   when='@:3.5.1',  # or until a version is released that is supported on non-x86_64 targets
>                    msg='ROCclr requires x86 intriniscs except on ARM')
>     # ...
> ```
> 
> though, ideally, the `ROCclr` CMakeLists wouldn't make assumptions about the architecture like this if it's possible to build it on Power ISA. I don't want to pollute this MR with the details, but I'd like to discuss the best approach to build the ROCm software stack from source on ppc64le/nvcc hardware. This includes
> 
> * what platform ("deprecated HCC" vs "ROCclr with x86 intrinics" vs "something I've missed") to target to build it successfully
> * and making clear what ROCm components are needed/useful on NVCC/non-x86_64 platforms through spack platform/target `conflicts()` constraints on the appropriate packages.
> 
> At a minimum, I'd like to offer users on our ppc64le/nvcc machines `hip` and `hipcc`. And while I know the math libs generally fallback to the respective CUDA counterparts/implementations when building HIP codes on an nvcc platform, our users have expressed interest in being able to use `hipBLAS`, `rocFFT`, `rocRAND`, `hipSPARSE` on our NVCC-platform machines. I think it would be a good idea to have an optional `cuda` variant and dependency on the libraries which need CUDA headers at build-time to support NVCC platforms.
> 
> To be clear, my interest here is to help our users currently running on Nvidia hardware have a headstart porting their codes to AMD/heterogeneous hardware. While it may seem counterproductive from certain perspectives to focus on NVCC platforms, the effort is ultimately so we can migrate users to AMD hardware with as much satisfaction and as little pain as possible.

Matt, 
Thank you for your detailed comments. I welcome your inputs. I see that you have a case for supporting NVCC and other platforms too. I would request to help me with the exact requirements so that we will address that in our future PRs. At this moment, i am okay to add the below to rocclr and other recipesIif valid).. 
   conflicts('target=ppc64',
                   when='@:3.5.1',  # or until a version is released that is supported on non-x86_64 targets
                   msg=' Valid Message')

haampie commented on var/spack/repos/builtin/packages/rocm-cmake/package.py:
```diff
@@ -0,0 +1,29 @@
+
+    variant('build_type', default='Release', values=("Release", "Debug"), description='CMake build type')
+
+    depends_on('cmake@3.5.2', type='build')
```
https://github.com/RadeonOpenCompute/rocm-cmake/blob/master/CMakeLists.txt#L5 here it says "cmake_minimum_required (VERSION 2.8) # TODO: Should probably be 3.0", which is a lowerbound, maybe just replace this with `depends_on('cmake@3:', type='build')`?

Edit: I see this is already addressed elsewhere

srekolam replied:
I find this little tricky.. version inside cmakelists.txt files mention cmake_minimum_required( Versions) appears very different for each of the rocm components. I am seeing conflicts due to this when spack is trying to concretize spec Already requested spack on it

adamjstewart replied:
I mentioned this above, but it should be:
```suggestion
    depends_on('cmake@2.8:', type='build')
```
The colon at the end is very important, it means that 2.8 is the minimum required version, but there is no maximum required version, it's an open range.

haampie commented on var/spack/repos/builtin/packages/llvm-amdgpu/package.py:
```diff
@@ -0,0 +1,35 @@
+    variant('build_type', default='Release', values=("Release", "Debug"), description='CMake build type')
+
+    depends_on('cmake@3.5.2', type='build')
+
```
```suggestion
depends_on('python', type='build')
```

srekolam replied:
@haampie , Will look into it..I was able to build it with out explicitly adding it .Is there a specific need to include it.. Appreciate your inputs.

adamjstewart replied:
If the package requires Python at build time, you should always add an explicit dependency, even if you happen to have Python installed on your OS.

haampie replied:
I have just added it because AUR seems to do it as well: https://aur.archlinux.org/packages/llvm-amdgpu/ and the llvm package seems to list it as a build dep as well

haampie commented on var/spack/repos/builtin/packages/hsakmt-roct/package.py:
```diff
@@ -0,0 +1,34 @@
+    variant('build_type', default='Release', values=("Release", "Debug"), description='CMake build type')
+
+    depends_on('cmake@3.5.2', type='build')
+
```
```suggestion
depends_on('numactl')
```
Otherwise `#include <numa.h>` will error

srekolam replied:
we have tried  including this already but it still requires numactl-libs and numactl-devel packages. Will work with spack on this open item

haampie commented on var/spack/repos/builtin/packages/comgr/package.py:
```diff
@@ -0,0 +1,36 @@
+    depends_on('rocm-cmake@3.5:', type='build', when='@3.5:')
+    depends_on('llvm-amdgpu@3.5:', type='build', when='@3.5:')
+    depends_on('rocm-device-libs@3.5:', type='build', when='@3.5:')
+
```
Theu AUR package has a dependency on `zlib` here.

haampie commented on var/spack/repos/builtin/packages/hsakmt-roct/package.py:
```diff
@@ -0,0 +1,34 @@
+        args = [
+            '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH="FALSE"',
+            '-DBUILD_SHARED_LIBS="on"',
+            '-DHSAKMT_WERROR=1'
```
```suggestion

```
Breaks compilation in GCC 9, see https://github.com/numactl/numactl/pull/26. It was never resolved upstream. Better to just not fail on errors.

srekolam replied:
@haampie , can you try yum install  numactl-devel  and numactl-libs package for hsakmt-roct.  The spack package for libnuma doesnot support the current needs for rocm. We plan to request spack team on adding support for these packages. If that is addressed the hsakmt-roct will be built sucessfuly

haampie replied:
What do you mean by 'does not support the current needs for rocm'? It seems to work fine with my suggested change

srekolam replied:
@haampie , I plan to include depends_on('numactl@2.0.12', type='link') and edit -DHSAKMT_WERROR=1 to -DHSAKMT_WERROR=0 . today.

haampie replied:
@srekolam you might be interested in many of my patches, i can submit a PR to your branch

haampie replied:
See https://github.com/srekolam/spack/pull/1 for my changes that solve many issues here

haampie commented:
With a bunch of changes I can almost get `spack install rocfft` to work, but not entirely.

One problem is that `hipcc` makes all kinds of assumptions about where things are installed:

```
[ 57%] Building CXX object library/src/device/CMakeFiles/rocfft-device.dir/kernel_launch_single_5.cpp.o
cd /tmp/harmen/spack-stage/spack-stage-rocfft-3.5.0-xgr7rtojamk33aupjy7pypjk6actxyha/spack-build/library/src/device && /home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-9.3.0/hip-3.5.0-utrxlkkpqii22ltt7f4klevwzcb7aqtg/bin/hipcc   -Drocfft_device_EXPORTS -I/tmp/harmen/spack-stage/spack-stage-rocfft-3.5.0-xgr7rtojamk33aupjy7pypjk6actxyha/spack-src/library/src/device -I/tmp/harmen/spack-stage/spack-stage-rocfft-3.5.0-xgr7rtojamk33aupjy7pypjk6actxyha/spack-src/library/src/device/../include -I/tmp/harmen/spack-stage/spack-stage-rocfft-3.5.0-xgr7rtojamk33aupjy7pypjk6actxyha/spack-src/library/src/device/kernels -I/tmp/harmen/spack-stage/spack-stage-rocfft-3.5.0-xgr7rtojamk33aupjy7pypjk6actxyha/spack-src/library/src/device/generator -I/tmp/harmen/spack-stage/spack-stage-rocfft-3.5.0-xgr7rtojamk33aupjy7pypjk6actxyha/spack-build/library/src/device -I/tmp/harmen/spack-stage/spack-stage-rocfft-3.5.0-xgr7rtojamk33aupjy7pypjk6actxyha/spack-src/library/include -I/tmp/harmen/spack-stage/spack-stage-rocfft-3.5.0-xgr7rtojamk33aupjy7pypjk6actxyha/spack-build/include  -O3 -DNDEBUG -fPIC   -fno-gpu-rdc -std=c++14 -o CMakeFiles/rocfft-device.dir/kernel_launch_single_5.cpp.o -c /tmp/harmen/spack-stage/spack-stage-rocfft-3.5.0-xgr7rtojamk33aupjy7pypjk6actxyha/spack-build/library/src/device/kernel_launch_single_5.cpp
Can't exec "/opt/rocm/bin/rocm_agent_enumerator": No such file or directory at /home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-9.3.0/hip-3.5.0-utrxlkkpqii22ltt7f4klevwzcb7aqtg/bin/hipcc line 686.
Use of uninitialized value $targetsStr in substitution (s///) at /home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-9.3.0/hip-3.5.0-utrxlkkpqii22ltt7f4klevwzcb7aqtg/bin/hipcc line 687.
Use of uninitialized value $targetsStr in split at /home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-9.3.0/hip-3.5.0-utrxlkkpqii22ltt7f4klevwzcb7aqtg/bin/hipcc line 693.
```

I think two things should happen:
- File an issue upstream about hipcc making too many assumption about installation locations. Basically they assume everything has a shared root `$ROCM_PATH`, from which they can access `$ROCM_PATH/lib`, `$ROCM_PATH/llvm/bin`, `$ROCM_PATH/hsa`, `$ROCM_PATH/hcc`, `$ROCM_PATH/bin/rocm_agent_enumerator`. Some of these can values can be overridden by environment variables, but not all.
- Add a required variant `amdgpu_target` which sets either a variable `HCC_AMDGPU_TARGET` or  `'-DCMAKE_CXX_FLAGS=--amdgpu-target={0}'.format(...)` to work around the error above, and set the `DEVICE_LIB_PATH`, `HIP_CLANG_PATH`, `HSA_PATH` variables, since they all otherwise depend on `ROCM_PATH`

haampie committed:
Use cmake 3.x instead of a fixed version

haampie committed:
Add a patch for cmake 3.x suport

haampie committed:
Fix rpath issue where hsa-rocr-dev does not allow it to be filled in by spack

haampie committed:
Remove inherited cmake args from comgr

haampie committed:
Make hsakmt-roct compile: no -Werror because with const cast in numa, and actually add the numa dependency

haampie committed:
Remove redundant cmake args which is inherited

haampie committed:
Fix some dependencies

haampie committed:
Fix some python 2.x compatibilities

haampie committed:
Add amd gpu targets to rocfft

haampie commented on var/spack/repos/builtin/packages/rocm-dbgapi/package.py:
```diff
@@ -0,0 +1,35 @@
+
+    depends_on('cmake@3.5.2', type='build')
+    depends_on('hsa-rocr-dev@3.5.0:', type='build', when='@3.5.0:')
+    depends_on('comgr@3.5.0:', type='build', when='@3.5.0')
```
This is a link dependency too

srekolam replied:
@haampie , @adamjstewart  , let me know if  you plan to merge the changes.

haampie committed:
Make comgr a link dep of rocm-dbgapi and remove redundant cmake args

haampie committed:
Remove redundant cmake args

haampie committed:
Remove more redundant cmake args

haampie committed:
Final redundant args

haampie committed:
Remove random variable

haampie committed:
Use installed rocclr instead of nonexisting directory

haampie committed:
Don't build outside the staging folder

haampie committed:
Deploy some missing cmake target file

haampie committed:
Formatting

haampie committed:
Fix target list

haampie committed:
Properly handle the rocclr dependency

haampie committed:
Formatting

haampie committed:
Fix vermin test

haampie commented:
I've incorporated many changes in https://github.com/spack/spack/pull/17898, which would resolve most issues from this PR

haampie committed:
Make all 3.5.0 package depend exactly on eachother

haampie committed:
Add a few missing link dependencies

haampie committed:
Fix flake8

haampie committed:
Remove some other redundant flags

haampie committed:
Add gcc install prefix for gcc builds of llvm-amdgpu

srekolam committed:
Merge pull request #1 from haampie/pull_17422-2

spack fixes

srekolam committed:
review changes for the spack recipes

haampie committed:
Do not hard-code versions

haampie committed:
Fix atmi install

- no more relative rpaths outside of install directory (required patch)
- fix build -> link dependencies
- remove unused build dependency

haampie commented:
@srekolam: the atmi package was still broken, it should link correctly when you merge https://github.com/srekolam/spack/pull/2. Also you still had a hard-coded and incorrect version number somewhere.

haampie committed:
Fix flake8 errors

haampie committed:
Remove unused variable and make things python 2.x compatible

haampie committed:
Fix flake8

haampie committed:
Move compiler config from rocfft -> hipcc

haampie committed:
Remove redundant dependency on fftw-api

haampie committed:
Remove redundant import

haampie committed:
Avoid hitting the ROCM_PATH variable altogether with a patch; also just fill in all variables

haampie committed:
Add missing deps z3, zlib and ncurses+termlib to llvm-amdgpu

haampie committed:
Fix perl shebang and add dep

srekolam committed:
Merge pull request #2 from haampie/pull_17422-2

More fixes

haampie committed:
Fix typo and patch HIP_CLANG_ROOT detection in hip's cmake files

haampie commented:
@srekolam can you merge https://github.com/srekolam/spack/pull/3 as well? It fixes a typo and adds a patch for hipcc which is required when using the hip package's cmake files.

srekolam committed:
Merge pull request #3 from haampie/pull_17422-2

Fix typo and patch HIP_CLANG_ROOT detection in hip's cmake files

adamjstewart commented on var/spack/repos/builtin/packages/rocm-smi-lib/package.py:
```diff
@@ -0,0 +1,30 @@
+
+    @run_after('install')
+    def post_install(self):
+        popen('cp -R {0}/rocm_smi/lib {1}'.format(self.prefix, self.prefix))
+        popen('cp -R {0}/rocm_smi/include {1}'.format(self.prefix,
+                                                      self.prefix))
+        popen('rm -R {0}/rocm_smi'.format(self.prefix))
```
These should use Spack and/or Python commands, not OS commands. We want Spack to someday support Windows, and there are a lot of differences in the flags that system commands use on macOS vs Linux.

Spack has a `copy_tree` and `install_tree` commands you can use to copy directories. As for deletion, I think the `os` module has some functions for that that work on any OS.

adamjstewart reviewed (commented):
Gave things another review. Most of the packages look good from a Spack perspective now, just a few last requests that still need to be addressed.

haampie commented:
Thanks @adamjstewart :) 

We are going to need https://github.com/spack/spack/pull/17986 for llvm-amdgpu too, otherwise rocBLAS will link to the system zlib and ncurses.

That would be https://github.com/srekolam/spack/pull/4. It's not strictly necessary for packages in this specific PR, but we got to have it anyways.

srekolam committed:
fixing build failure due z3 and adding zlib for rocgdb

haampie commented on var/spack/repos/builtin/packages/llvm-amdgpu/package.py:
```diff
@@ -22,9 +22,7 @@ class LlvmAmdgpu(CMakePackage):
-    depends_on("z3")
-    depends_on("zlib")
-    depends_on("ncurses+termlib")
+    depends_on('zlib', type='link')
```
Can you please motivate why you made this change?

Note from https://github.com/RadeonOpenCompute/llvm-project/blob/amd-stg-open/llvm/CMakeLists.txt#L349:

```
option(LLVM_ENABLE_TERMINFO "Use terminfo database if available." ON)
```
ncurses/terminfo is enabled by default, and you just removed the dependency.

And z3 support is automatically enabled when z3 is found. `llvm-amdgpu` compiles fine for me without your changes.

srekolam replied:
@haampie , basically i am seeing a build failure due to z3 while building rocm-device-libs.  llvm-amdgpu will build fine with the above inclusions but rocm-device-libs is failing with the below -
>> 62    /opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/ld:
           cannot find -ltinfo
  >> 63    collect2: error: ld returned 1 exit status
  >> 64    make[2]: *** [utils/prepare-builtins/prepare-builtins] Error 1
     65    make[2]: Leaving directory `/tmp/root/spack-stage/spack-stage-rocm-d
           evice-libs-3.5.0-dxvh2xe47levmc3hmdedj62niaephda5/spack-build'
  >> 66    make[1]: *** [utils/prepare-builtins/CMakeFiles/prepare-builtins.dir
           /all] Error 2
     67    make[1]: Leaving directory `/tmp/root/spack-stage/spack-stage-rocm-d
           evice-libs-3.5.0-dxvh2xe47levmc3hmdedj62niaephda5/spack-build'
  >> 68    make: *** [all] Error 2

haampie replied:
`cannot find -ltinfo` is not `z3`, it's `ncurses`. Possibly related is https://github.com/srekolam/spack/pull/4, can you share the full linker command?

haampie replied:
Indeed it is... https://github.com/RadeonOpenCompute/ROCm-Device-Libs/blob/amd-stg-open/utils/prepare-builtins/CMakeLists.txt#L29 it's using the llvm support target.

haampie replied:
@srekolam can you merge https://github.com/srekolam/spack/pull/4 and try again where you have the `z3`, `zlib` and `ncurses` dependencies?

srekolam replied:
@haampie ,i got the patch and applied the changes manually. I see the issue is resolved. the rocm-device-libs builds with out any errors.

haampie commented on var/spack/repos/builtin/packages/rocm-device-libs/package.py:
```diff
@@ -0,0 +1,33 @@
+
+    depends_on('cmake@3:', type='build')
+    for ver in ['3.5.0']:
+        depends_on('llvm-amdgpu@' + ver, type='build', when='@' + ver)
```
You need to make this a link dependency too, so just write:
```suggestion
        depends_on('llvm-amdgpu@' + ver, type='build', when='@' + ver)
        depends_on('llvm-amdgpu@' + ver, when='@' + ver)
```

srekolam committed:
new changes to add z3,curses dependency for llvm-amdgpu

srekolam committed:
Merge remote-tracking branch 'upstream/develop' into develop

srekolam committed:
fix flake8 error

srekolam commented:
@adamjstewart , I have updated recipes with all the review comments. I am planning a follow up PR after this and will address rocm-smi-lib and rocm-smi recipes ( usage of cp -R ) along with other recipes.

adamjstewart reviewed (approved):
I think there's still a lot of work that can be done to make these packages more robust, but that can wait for smaller follow-up PRs. It seems like there is a lot of demand for these packages in Spack, so I'm going to merge this now and we can iterate and smaller bug fixes later. Thanks for everyone's hard work!

srekolam commented:
@adamjstewart  @haampie  , Thanks for all the inputs.
